### DISCUSSION TITLE
`Invalid source name` when trying to import openzeppelin inside a monorepo

### DISCUSSION THREAD

**Discussion by SebastienGllmt** [NomicFoundation/hardhat#4729] [Troubleshooting] [answered]
# Goal

use `@openzeppelin/contracts` within a package in my monorepo

# Error

I have a monorepo, and one of my packages is my evm contracts folder
```
root
    - node_modules
    - packages
        - evm-contracts
            - hardhat.config.ts
            - package.json
        - app
```

I want to use `@openzeppelin/contracts` from inside my package and use `foundry` so I've done the following:
1. I installed `@openzeppelin/contracts` and `@openzeppelin/contracts-upgradeable`
2. I installed `@nomicfoundation/hardhat-foundry`

Now, since dependencies for my project are hoisted to the root of the repository, I've set the following foundry configuration: `@openzeppelin/=../../../node_modules/@openzeppelin/`

This works and now foundry compiles. However, when I run `npx hardhat compile` I get the following error

```
HardhatError: HH1001: Invalid source name ../../../node_modules/@openzeppelin/contracts/token/ERC20/ERC20.sol. Expected source name but found a relative path.
    at validateSourceNameFormat (my-project/node_modules/hardhat/src/utils/source-names.ts:27:11)
    at Resolver.resolveSourceName (my-project/node_modules/hardhat/src/internal/solidity/resolver.ts:97:15)
    at Resolver.resolveImport (my-project/node_modules/hardhat/src/internal/solidity/resolver.ts:197:35)
    at async my-project/node_modules/hardhat/src/internal/solidity/dependencyGraph.ts:188:28
    at async Promise.all (index 0)
    at async DependencyGraph._addDependenciesFrom (my-project/node_modules/hardhat/src/internal/solidity/dependencyGraph.ts:186:5)
    at async Promise.all (index 0)
    at async Function.createFromResolvedFiles (my-project/node_modules/hardhat/src/internal/solidity/dependencyGraph.ts:15:5)
    at async Environment._runTaskDefinition (my-project/node_modules/hardhat/src/internal/core/runtime-environment.ts:358:14)
    at async Environment.run (my-project/node_modules/hardhat/src/internal/core/runtime-environment.ts:191:14)
```

# Debugging

I tried to debug the [resolveSourceName](https://github.com/NomicFoundation/hardhat/blob/7630646422deb62f5d082ff4d80c0271ea87d5f8/packages/hardhat-core/src/internal/solidity/resolver.ts#L85) function in `hardhat-core` and found the following:
1. The input to the function is `sourceName: "@openzeppelin/contracts/token/ERC20/ERC20.sol"`
2. This gets updated to `remappedSourceName: "../../../node_modules/@openzeppelin/contracts/token/ERC20/ERC20.sol"`
3. It then crashes on the `validateSourceNameFormat` call in this function

Notably, this fails because [validateSourceNameFormat explicitly disallows relative paths](https://github.com/NomicFoundation/hardhat/blob/7630646422deb62f5d082ff4d80c0271ea87d5f8/packages/hardhat-core/src/utils/source-names.ts#L26)

# Question

If relative paths are explicitly disabled, is there another way I can use `@openzeppelin/contracts` together with `hardhat` within my monorepo?

To be clear, removing `import '@nomicfoundation/hardhat-foundry';` in my hardhat config solves the issue.

Possibly related: https://github.com/NomicFoundation/hardhat/issues/4194 mentioned the remapping not being detected. In my case, it's definitely being detected (but the fact it's being used is breaking things)

**Comment by fvictorio** [comment, marked as answer]
Thank you @SebastienGllmt, I think this merits being an issue, so I opened https://github.com/NomicFoundation/hardhat/issues/4734 to look into it next time we iterate the hardhat-foundry plugin.
